### PR TITLE
Qemu CmpLog

### PR DESCRIPTION
**Pull request by andreafioraldi**

### CODE REVIEW

File: libafl_qemu/src/executor.rs [hunk 1]
```diff
@@ -67,27 +76,61 @@ where
...
     }
 
     #[allow(clippy::unused_self)]
-    pub fn hook_block_execution(&self, hook: extern "C" fn(u64)) {
+    pub fn hook_block_execution(&self, hook: extern "C" fn(addr: u64)) {
         emu::set_exec_block_hook(hook);
     }
+
+    #[allow(clippy::unused_self)]
+    pub fn hook_cmp_generation(&self, hook: fn(&mut S, addr: u64, size: usize) -> Option<u32>) {
```
Comment (domenukk) [inline libafl_qemu/src/executor.rs:110]: Why not leave the `&self`? One less argument passed. Shouldn't matter fro LTO, but is there a benefit?

**Review by domenukk — APPROVED** [review]
Beautiful

File: fuzzers/fuzzbench_qemu/Makefile [hunk 2]
```diff
@@ -33,6 +33,7 @@ target/$(BUILD_TARGET)/lib$(FUZZER_NAME).a: src/*
 
 qemu-libafl-bridge:
 	git clone git@github.com:AFLplusplus/qemu-libafl-bridge.git
+	cd qemu-libafl-bridge && git checkout f71558480c336b51360e8d382b95c2bb753b785b
```
[the remark this answers is not in the document]
Comment (andreafioraldi) [reply fuzzers/fuzzbench_qemu/Makefile:36]: Nope if we are going to have more than 1 qemu fuzzer a submodule is not ok imo

File: libafl_qemu/src/executor.rs (hunk 1 above)
Comment (andreafioraldi) [reply libafl_qemu/src/executor.rs:110]: I need self for the generic

File: libafl_qemu/src/hooks.rs [hunk 3]
```diff
@@ -32,17 +55,56 @@ where
         .metadata_mut()
         .get_mut::<QemuEdgesMapMetadata>()
         .unwrap();
-    Some(*meta.map.entry((src, dest)).or_insert_with(|| unsafe {
-        let id = MAX_EDGES_NUM;
-        MAX_EDGES_NUM = (MAX_EDGES_NUM + 1) & (EDGES_MAP_SIZE - 1);
-        id as u32
-    }))
+    let id = max(meta.current_id as usize, unsafe { MAX_EDGES_NUM });
+    if !meta.map.contains_key(&(src, dest)) {
```
[the remark this answers is not in the document]
Comment (andreafioraldi) [reply libafl_qemu/src/hooks.rs:59]: In release mode it is the same imo, get cannot be used as I have to do side effect to current_id while already borrowing the meta as mut

File: libafl_qemu/src/executor.rs (hunk 1 above)
Comment (domenukk) [reply libafl_qemu/src/executor.rs:110]: Makes sense thx

File: fuzzers/fuzzbench_qemu/Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (domenukk) [reply fuzzers/fuzzbench_qemu/Makefile:36]: how about pulling it in build.rs inside libafl_qemu instad of manually?
Not relevant for this PR I guess

**Pull request merged**